docs/en/concepts-and-designs/lal.md

```diff
@@ -0,0 +1,239 @@
+
+```
+
+The other metrics generated is `http_response_time`, so that you can configure MAL rules to generate more useful metrics like percentiles.
+
+```yaml
+# ... other configurations of MAL
+
+metrics:
+  - name: response_time_percentile
+    exp: http_response_time.sum(['le', 'service', 'instance']).increase('PT5M').histogram().histogram_percentile([50,70,90,99])
+```
+
+### Sink
+
+Sinks are the persistent layer of the LAL. By default, all the logs of each filter are persisted into the storage. However, there are some mechanisms that allow you to selectively save some logs, or even drop all the logs after you've extracted useful information, such as metrics.
+
+#### Sampler
+
+Sampler allows you to save the logs in a sampling manner. Currently, 2 sampling strategies are supported, `ratelimit` and `probabilistic`. If multiple samplers are specified, the last one determines the final sampling result, see examples in [Enforcer](#enforcer).
+
+`ratelimit` samples `n` logs at most, in a given duration (e.g. 1 second).
+`probabilistic` samples `n%` logs .
+
+Examples:
+
+```groovy
+filter {
+    // ... parser
+    
+    sink {
+        sampler {
+            ratelimit 100.per.second // 100 logs per second
+        }
+    }
+}
+```
+
+```groovy
+filter {
+    // ... parser
+    
+    sink {
+        sampler {
+            probabilistic 50.percent // 50% logs
+        }
+    }
+}
+```
+
+#### Dropper
+
+Dropper is a special sink, meaning that all the logs are dropped without any exception. This is useful when you want to drop debugging logs,
+
+```groovy
+filter {
+    // ... parser
+    
+    sink {
+        if (parsed.level == "DEBUG") {
+            dropper {}
+        } else {
+            sampler {
+                // ... configs
+            }
+        }
+    }
+}
+```
+
+or you have multiple filters, some of which are for extracting metrics, only one of them needs to be persisted.
+
+```groovy
+filter { // filter A: this is for persistence
+    // ... parser
+
+    sink {
+        sampler {
+            // .. sampler configs
+        }
+    }
+}
+filter { // filter B:
+    // ... extractors to generate many metrics
+    extractors {
+        metrics {
+            // ... metrics
+        }
+    }
+    sink {
+        dropper {} // drop all logs because they have been saved in "filter A" above.
+    }
+}
+```
+
+#### Enforcer
+
+Enforcer is another special sink that forcibly samples the log, a typical use case of enforcer is when you have configured a sampler and want to save some logs forcibly, for example, to save error logs even if the sampling mechanism is configured.
+
+```groovy
+filter {
+    // ... parser
+    
+    sink {
+        sampler {
+            // ... sampler configs
+        }
+        if (parserd["level"] == "ERROR") { // sample error logs even if the sampling strategy is configured
+            enforcer {
+            }
+        }
+    }
```

Comment:
I am a little confused about this example, if the data run as a flow, then, no sampled data should be ignored already. Is this meaning, all codes will execute, and sampled is only a flag rather than a `return statement`?

Reply:
> Is this meaning, all codes will execute, and sampled is only a flag rather than a `return statement`?

Right, sampling result is a flag, only when the persistence really happens, the sampling result takes effect

Reply:
Got it.